src/Subject.ts

```diff
@@ -2,236 +2,145 @@ import {Operator} from './Operator';
+    } else {
+      this.observers.push(subscriber);
+      return new SubjectSubscription(this, subscriber);
     }
-
-    this._complete();
   }
 
   asObservable(): Observable<T> {
-    const observable = new SubjectObservable(this);
+    const observable = new Observable<T>();
+    (<any>observable).source = this; //HACKITY HACK
     return observable;
   }
+}
 
-  protected _next(value: T): void {
-    if (this.destination) {
-      this.destination.next(value);
-    } else {
-      this._finalNext(value);
-    }
+/**
+ * @class AnonymousSubject<T>
+ */
+export class AnonymousSubject<T> extends Subject<T> {
+  constructor(protected destination?: Observer<T>, protected source?: Observable<T>) {
+    super();
   }
 
-  protected _finalNext(value: T): void {
-    let index = -1;
-    const observers = this.observers.slice(0);
-    const len = observers.length;
-
-    while (++index < len) {
-      observers[index].next(value);
+  next(value: T) {
+    const { destination } = this;
+    if (destination && destination.next) {
+      destination.next(value);
     }
   }
 
-  protected _error(err: any): void {
-    if (this.destination) {
+  error(err: any) {
+    const { destination } = this;
+    if (destination && destination.error) {
       this.destination.error(err);
-    } else {
-      this._finalError(err);
-    }
-  }
-
-  protected _finalError(err: any): void {
-    let index = -1;
-    const observers = this.observers;
-
-    // optimization to block our SubjectSubscriptions from
-    // splicing themselves out of the observers list one by one.
-    this.observers = null;
-    this.isUnsubscribed = true;
-
-    if (observers) {
-      const len = observers.length;
-      while (++index < len) {
-        observers[index].error(err);
-      }
     }
-
-    this.isUnsubscribed = false;
-
-    this.unsubscribe();
   }
 
-  protected _complete(): void {
-    if (this.destination) {
+  complete() {
+    const { destination } = this;
+    if (destination && destination.complete) {
       this.destination.complete();
-    } else {
-      this._finalComplete();
-    }
-  }
-
-  protected _finalComplete(): void {
-    let index = -1;
-    const observers = this.observers;
-
-    // optimization to block our SubjectSubscriptions from
-    // splicing themselves out of the observers list one by one.
-    this.observers = null;
-    this.isUnsubscribed = true;
-
-    if (observers) {
-      const len = observers.length;
-      while (++index < len) {
-        observers[index].complete();
-      }
     }
-
-    this.isUnsubscribed = false;
-
-    this.unsubscribe();
   }
 
-  private throwIfUnsubscribed(): void {
-    if (this.isUnsubscribed) {
-      throwError(new ObjectUnsubscribedError());
+  _subscribe(subscriber: Subscriber<T>): Subscription {
+    const { source } = this;
+    if (source) {
+      return this.source.subscribe(subscriber);
+    } else {
+      return Subscription.EMPTY;
     }
   }
-
-  [$$rxSubscriber]() {
```

Comment:
No, the only place it was used was changed because we have to wrap Subjects in a SafeSubscriber to prevent them from being `unsubscribed` and disabled (via throwing `ObjectUnsubscribedErrors`) in cases where they complete or error. This was done to match Rx4